retrieval/discovery/ecs.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	ecs "github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"golang.org/x/net/context"
+	"time"
+)
+
+// ECSDiscovery periodically performs ECS-SD requests. It implements
+// the TargetProvider interface.
+type ECSDiscovery struct {
+	aws         *aws.Config
+	interval    time.Duration
+	portFrom    int64
+	portTo      int64
+	clusterName string
+}
+
+// NewECSDiscovery returns a new ECSDiscovery which periodically refreshes its targets.
+func NewECSDiscovery(conf *config.ECSSDConfig) *ECSDiscovery {
+	awsConfig := defaults.Get().Config
+
+	// override the default region
+	if conf.Region != "" {
+		awsConfig.Region = &conf.Region
+	}
+	if conf.AccessKey != "" && conf.SecretKey != "" {
+		creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+		awsConfig.Credentials = creds
+	}
+
+	return &ECSDiscovery{
+		aws:         awsConfig,
+		interval:    time.Duration(conf.RefreshInterval),
+		portFrom:    int64(conf.PortFrom),
+		portTo:      int64(conf.PortTo),
+		clusterName: conf.ClusterName,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *ECSDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := ed.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ed.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ed *ECSDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	ecsService := ecs.New(session.New(), ed.aws)
+	ec2s := ec2.New(session.New(), ed.aws)
+
+	listServicesResponse, err := ecsService.ListServices(&ecs.ListServicesInput{
+		Cluster:    aws.String(ed.clusterName),
+		MaxResults: aws.Int64(100),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing ecs services: %s", err)
+	}
+
+	services, err := ecsService.DescribeServices(&ecs.DescribeServicesInput{
+		Cluster:  aws.String("default"),
+		Services: listServicesResponse.ServiceArns,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing describing services: %s", err)
+	}
+
+	for _, service := range services.Services {
+		// Possibly cache theses, since ARNS are immutable, afaik
+		taskDefinitionResponse, err := ecsService.DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
+			TaskDefinition: service.TaskDefinition,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error listing describing task description: %s", err)
+		}
+
+		monitored := false
+		monitoringPort := 0
+
+		var dockerLabels map[string]*string
+
+		// See if any container is exposing the specified ports
```

Comment:
say that you say that port 8123 on the container is the one exposing the /metrics handler by specifying port: 8123 in the config. Once we see that port being mapped to some port on the host, which it needs to be to be accessible, we will use the port which it is mapped to. 

So in your config, you specify the port which is exposed on the container, and by reading the ECS configuration, we figure out which port that is mapped to on the host. 

I prefer that way, since the the user can set this up without having to deal with remapping it all. Just by specifying the container port on the config.